test/e2e/operator_expectations.go

```diff
@@ -608,3 +613,481 @@ func (tc *testConfig) ExpectAutoscalerScalesOut() error {
 		return len(nodeList.Items) == clusterInitialTotalNodes, nil
 	})
 }
+
+func (tc *testConfig) ExpectNodeToBeDrainedBeforeDeletingMachine() error {
+	listOptions := client.ListOptions{
+		Namespace: namespace,
+	}
+
+	var machine mapiv1beta1.Machine
+	var nodeName string
+	var node *corev1.Node
+
+	glog.Info("Get machineList with at least one machine with NodeRef set")
+	if err := wait.PollImmediate(1*time.Second, waitShort, func() (bool, error) {
+		machineList := mapiv1beta1.MachineList{}
+		if err := tc.client.List(context.TODO(), &listOptions, &machineList); err != nil {
+			glog.Errorf("error querying api for machineList object: %v, retrying...", err)
+			return false, nil
+		}
+		for _, machineItem := range machineList.Items {
+			// empty or non-worker role skipped
+			if machineItem.Labels["sigs.k8s.io/cluster-api-machine-role"] == "worker" {
+				if machineItem.Status.NodeRef != nil && machineItem.Status.NodeRef.Name != "" {
+					machine = machineItem
+					nodeName = machineItem.Status.NodeRef.Name
+					return true, nil
+				}
+			}
+		}
+		return false, fmt.Errorf("no machine found with NodeRef not set")
+	}); err != nil {
+		return err
+	}
+
+	glog.Info("Get nodeList")
+	if err := wait.PollImmediate(1*time.Second, waitShort, func() (bool, error) {
+		nodeList := corev1.NodeList{}
+		if err := tc.client.List(context.TODO(), &listOptions, &nodeList); err != nil {
+			glog.Errorf("error querying api for nodeList object: %v, retrying...", err)
+			return false, nil
+		}
+		for _, nodeItem := range nodeList.Items {
+			if nodeItem.Name == nodeName {
+				node = &nodeItem
+				break
+			}
+		}
+		if node == nil {
+			return false, fmt.Errorf("node %q not found", nodeName)
+		}
+		return true, nil
+	}); err != nil {
+		return err
+	}
+
+	glog.Info("Delete machine and observe node draining")
+	if err := tc.client.Delete(context.TODO(), &machine); err != nil {
+		return fmt.Errorf("unable to delete machine %q", machine.Name)
+	}
+
+	return wait.PollImmediate(time.Second, waitShort, func() (bool, error) {
+		eventList := corev1.EventList{}
+		if err := tc.client.List(context.TODO(), &listOptions, &eventList); err != nil {
+			glog.Errorf("error querying api for eventList object: %v, retrying...", err)
+			return false, nil
+		}
+
+		glog.Infof("Fetching delete machine and node drained events")
+		var nodeDrainedEvent *corev1.Event
+		var machineDeletedEvent *corev1.Event
+		for _, eventItem := range eventList.Items {
+			if eventItem.Reason == "Deleted" && eventItem.Message == fmt.Sprintf("Node %q drained", nodeName) {
+				nodeDrainedEvent = &eventItem
```

Comment:
We're taking the address of the local loop variable here - is this intentional?

Reply:
It is not, it should be changed actually. Though, `ExpectNodeToBeDrainedBeforeDeletingMachine` was superseded by `ExpectNodeToBeDrainedBeforeMachineIsDeleted` so it can be deleted.

Reply:
Thanks! Seems ok since this is not storing appended values or similar
This is just bringing @ingvagabund draining tests from aws. To validate the new rbac does not break current coverage

Right after this we can merge https://github.com/openshift/machine-api-operator/pull/219
and https://github.com/openshift/cluster-api-actuator-pkg/pull/28
And start rewriting the e2e suite there addressing this specific concern and others